WIP swap callmemaybe with etcd

SomeoneToIgnore opened the pull request.

hlinnaka commented on libs/etcd_broker/src/lib.rs:
```diff
@@ -0,0 +1,288 @@
+            format!("Failed to init the watch for subscription {subscription_kind:?}")
+        })?;
+
+    let (timeline_updates_sender, safekeeper_timeline_updates) = mpsc::unbounded_channel();
```
I don't understand the purpose of this channel. Let me try to explain how I thought this would work:

The WAL receiver wants to wake up, when the state in etcd changes. To do that, we create an etcd watch. The watch is a future that returns when something changed for the timeline we're interested in.

In the WAL receiver loop, we simultenously wait for the etcd watch, and for the I/O on the libpq connection.

What do we need the channel for?

SomeoneToIgnore replied:
I've pushed a few more docs that aim to describe the current impl state, but indeed I find the current approach a bit hard too, so I'd love to refactor it more before adding any proper docs.

The

> we simultenously wait for the etcd watch

is the one that I'm not sure how to implement easier: I'm aware of the following tokio future manipulation primitives:
* `.await` and nothing after this in the block can be executed
* `tokio::select!` that runs multiple futures and processes the one that's run first, "cancelling" the rest
There's a ["cancellation safety"](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) notion that limits what could safely be placed in such `select!` and I'm not sure that wal streaming with all the buffers is cancellation safe, since the tasks get abruptly shut down
* `tokio::join!`/`FuturesOrdered`/`FuturesUnordered` for concurrently running _all_ the tasks to the end.
* `tokio::spawn`/`tokio::task::spawn_blocking` for "run it on a separate thread" analogue, with similar Send + Sync + 'static requirements

So there are multiple async sources of data, each with its own cadence and potentially blocking behaviour:
* `tenant_mgr` populates its timeline map and notifies the main WAL receiver thread about this. Same for the removal
Handed in a top level wal receiver thread, since has to spawn timeline-wise broker loops
* `etcd` that's very swift to produce messages, minimum one per second, now handled in the broker loop
* the actual WAL stream from safekeeper, that could take a while to open, will hopefully stay open for a long time, if the safekeeper we're connected to is healthy

So it looks like if we place the WAL streaming task to `select!` with etcd and the rest, it will get cancelled too quickly due to etcd message storm.
If we place it to `join!`, then nothing gets updated before the entire wal stream loop finishes.
So I've moved away the WAL loop into a separate tokio task via `spawn!`, it's all the existing WAL streaming code basically, it was almost async already.
We need to react on the task finish, to restart a connection immediately hence one of the channels + the other one to cancel the entire WAL receiving (if we abort the connection).

etcd subscriptio + connection management lives one level up, starts per timeline and also needs some channel to signal that it has to stop: has to be different from the WAL stop signal, since broker and safekeeper connections are managed separately.

then, the top level watched of the channel that gives us new tenant tasks (add/remove), so that we could start or stop WAL streaming for those.


Unless I'm missing some other good way to concurrently poll tasks (without manual state machine writing), I think that the "nested control loops" structure will stay, but will get wrapped into more readable structs.
What I want to remove are the channels, and also investirage, if `tokio::task::JoinHandle` cancellation is a fine thing to do: then indeed we can reduce the number of channels by `select!` on join handles directly.

arssher replied:
My understanding is the same, i.e. that running communication with broker in separate task is the most sensible way.

> tokio::task::JoinHandle cancellation is a fine thing to do

From the common sense I suppose so, but don't see a confirmation in the docs, so better check.

SomeoneToIgnore replied:
Had worked on that more, there's a way to await on the handle as `(&mut handle).await` and it seems to behave ok and not to cancel the task after the future drop.

But it's not really needed with the current model, when I have to get periodic events from the WAL receiver task to ensure we keep the connection up-to-date.

hlinnaka commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -20,22 +19,27 @@ use std::collections::hash_map::Entry;
 
     lazy_static::lazy_static! {
         static ref TENANTS: RwLock<HashMap<ZTenantId, Tenant>> = RwLock::new(HashMap::new());
+        static ref TIMELINE_UPDATE_SENDER: RwLock<Option<mpsc::UnboundedSender<LocalTimelineUpdate>>> = RwLock::new(None);
```
What is this channel for? Comment please

hlinnaka reviewed (commented):
I'm a bit concerned about all the new async code. Previously, each WAL receiver was a separate thread, and it only used async, with a "current thread" tokio runtime to execute the parts of rust-postgres that have to be async. Now it seems that there's a lot more async stuff happening, with long-running spawned tasks. What exactly is the border between sync and async code now?

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -1,170 +1,567 @@
+    if let Some((new_safekeeper_id, new_safekeeper_timeline)) = safekeeper_timelines
+        .into_iter()
+        .max_by_key(|(_, info)| info.flush_lsn)
+        .filter(|(_, info)| info.flush_lsn > Some(timeline.tline.get_disk_consistent_lsn()))
```
To avoid idle walreceivers, there is also should be logic for shutdown: if safekeepers stopped posting data about timeline, it means timeline became inactive and walreceiver should be stopped. We need to take into accounts deletes from etcd (happening after lease expires, soon after absense of new pushes) and remember old state for that.

(or invent some simpler way)

SomeoneToIgnore replied:
Addressed and added some tests on top.
Tests are somewhat clumsy, but the whole thing seem to select safekeepers normally now.

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -1,170 +1,567 @@
+        ThreadKind::WalReceiver,
+        None,
+        None,
+        "WAL receiver main thread",
```
Let's name it differently, e.g. WAL receiver launcher thread (and other funcs like walreceiver_main_thread_loop_step), as it is not really walreceiver.

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -1,170 +1,567 @@
+                        };
+
+                    let mut connection_failed = true;
+                    for attempt in 0..max_retries.get() {
```
I don't think we need any tracking of failures here, on task start. Broker task can fail later at any moment (you'd just see watch channel closed at current coding), so that wouldn't help anyway. And currently no one ever restarts broker tasks, that's wrong. We need infinite reconnect-to-etcd logic either inside the task or in walreceiver_main_thread level. At some point I realized that sharing etcd client among tasks is good for performance (still need to check to be absolutely sure, but looks like they share underlying http/2 connection and request multiplexing), so the latter is preferrable.

However, problem with etcd shouldn't immediately kill the current walreceiver connection I guess, so it might be better to track walreceiver state in shared memory so that next broker task can pick it up.

SomeoneToIgnore replied:
Replaced with the infinite loop, no more killing on that thing is done.

I've not placed this particular state into some shared mem, since turned out that the state we need for the broker is easier to recreate, but could be improved later if needed.

We would want to improve the subscription management based on the recent discussions, so we keep a single conneciton for multiple subscriptions, but that requires more work on our `broker` crate and not done right now.

arssher reviewed (commented):
One thing also concerning me here is that walreceiver thread now is tokio task, not separate thread. I'm not familiar with pageserver code enough at the moment to say for sure what is better, but I think we either need some argumentation confirming it is okay or switch back to separate thread, it doesn't create any problems as far as I see.

SomeoneToIgnore commented:
I'm very close to fix all that I've wanted in this PR, so marking it as ready to review.
Things that are missing still:
* remove the `callmemaybe_connstring` from compute: can go in a separate PR after this is merged
* fix 2 failing tests. No idea how to do this now, any help is very much welcome

Will work on tests fixing and addressing any PR feedback, if it comes.

SomeoneToIgnore commented on pageserver/src/thread_mgr.rs:
```diff
@@ -91,7 +91,7 @@ pub enum ThreadKind {
     PageRequestHandler,
 
-    // Thread that connects to a safekeeper to fetch WAL for one timeline.
+    // Main walreciver thread that ensures that every timeline spanws a connection to safekeeper for WAL fecthing.
```
Yes, the role had changed, I've adjusted the docs around, but the situation is so now, that
* there's an only thread, the main one that looks for more events from `tenant_mgr` and the cancellation
* there's a bunch of tokio tasks, 2 per timeline, that are launched for watching etcd and WAL streaming correspondingly

Nothing else, besides some extra threads that are used during `tokio::spawn` (none now, since we use the one threaded runtime, but can be more if we switch to another flavour) and extra threads that are used during `tokio::task::spawn_blocking` (~2 per timeline init, short-lived)

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -7,164 +7,526 @@
+
+    let timeline = Arc::clone(timeline);
+    let broker_join_handle = tokio::spawn(async move {
+        info!("WAL receiver broker thread had started");
```
Yes, had commented on it https://github.com/neondatabase/neon/pull/1619#discussion_r878253690 , let's continue there if you want to still.

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -1,170 +1,567 @@
+    broker_prefix: &str,
+    etcd_endpoints: &[Url],
+    timeline_updates_receiver: &'a mut mpsc::UnboundedReceiver<LocalTimelineUpdate>,
+    local_timeline_wal_receivers: &'a mut HashMap<
```
Yes, this is a bit dirty place.
Not sure [de]activation should go through this module indeed, but turns out now that this is the only place which knows when a tenant has no WAL streaming timelines (or gets one), so I've left this code here.

I could have added another channel and put it into some other thread loop, but that does look like an overkill.

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -1,170 +1,567 @@
+                    info!("Loop cancelled, shutting down");
+                    break;
+                }
+                walreceiver_result = walreceiver_task_completion.changed() => {
```
Looks like we have to keep the channel, since it's the only way for us to understand, whether the WAL receiver thread/task got any updates from safekeeper: it's crucial for us to detect the lagging ones and not sure there'a a way around it.

We also have to share some public info about WAL streaming via HTTP API now, so not sure the `join` approach works now.

petuhovskiy reviewed (commented):
Looks good, I'd like to merge it sooner and live with it in staging for some time. I see that some tests are failing, I can try to look into them if you haven't found the causes yet.

SomeoneToIgnore commented on control_plane/src/compute.rs:
```diff
@@ -345,7 +345,6 @@ impl PostgresNode {
             // This isn't really a supported configuration, but can be useful for
             // testing.
             conf.append("synchronous_standby_names", "pageserver");
-            conf.append("zenith.callmemaybe_connstring", &self.connstr());
```
Yes, and I plan to remove the corresponding code in compute after this PR is merged and no calamities happen for a week or so.

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1264 @@ thread_local! {
+                    //
+                    // When a task is spawned, it will loop until the subscription is cancelled, retrieving the timeline updates and deciding on the WAL connection state.
+                    for attempt in 0.. {
+                        exponential_backoff(attempt, 2.0, 60.0).await;
```
>  Is that intended?

Not at all, good catch, thank you.

> can return error only if we have some problems with etcd?

Yes, that's the intention: without etcd connection, we're pretty useless anyway, so, if configured to use one, have not much choice but to try accessing it, unless I miss something.
I've tried to reflect it on a module level documentation: 
> //! This is done endlessly, to ensure we don't miss the WAL streaming when it gets available on one of the safekeepers.

SomeoneToIgnore commented:
I've fixed both issues pointed out by @petuhovskiy and looks like it fixed most of the issues, thank you a lot for spotting those:
* the extra timeout in `WalReceiverConnection::new` for handling the open connection has to be removed, a cancellation is added instead. This fixed all but `test_pageserver_catchup_while_compute_down` tests for me.
* the etcd resubscription in an endless loop moved into the etcd broker main loop task which fixed `test_pageserver_catchup_while_compute_down` **release** runs.

Now there's only one test failing only on a debug build, `test_pageserver_catchup_while_compute_down`, so I'll look at the timings to see what's wrong.

petuhovskiy reviewed (approved):
The code looks nice and works ok in `test_restarts_under_load`. I think the only thing left is to debug `neon-cloud-e2e` tests, and after that we can merge it.

SomeoneToIgnore commented:
Neon tests were also flacky, I've seen one job failing here and there, so let's merge the whole thing?

I'll check those and will wait for the evening for any contradictions.

petuhovskiy commented:
@SomeoneToIgnore okay, let's merge then

SomeoneToIgnore commented:
Looks like I've broken the e2e tests after all, looking. Weird though, they used to work yesterday on this branch without issues.

LizardWizzard reviewed (commented):
Looks really nice! Thanks. I've enjoyed the reading :)

Few concerns:
* During experiments I saw that a single walreceiver thread was using the whole CPU core, and here we use single-threaded runtime for all tenants which is concerning.
* Quirks with joining wal receiver activity during detach, see the comments

petuhovskiy commented:
> During experiments I saw that a single walreceiver thread was using the whole CPU core, and here we use single-threaded runtime for all tenants which is concerning.

That's a good point, we should have performance tests for walreceiver when we have multiple active timelines. Without performance tests it's hard to tell what will happen here.

SomeoneToIgnore commented on libs/etcd_broker/src/lib.rs:
```diff
@@ -55,7 +55,9 @@ pub struct SkTimelineInfo {
-    pub safekeeper_connection_string: Option<String>,
+    pub safekeeper_connstr: Option<String>,
+    #[serde(default)]
+    pub pageserver_connstr: Option<String>,
```
I just follow the path of least resistance and don't change the code that is in safekeeper already.
Still not sure what should be better, @\kelvich proposed to have a "connected to compute" bit here instead, there might be better ideas on how to communicate pageserver's `NodeId` now that it has one.

Not sure we want to change it now, though?

LizardWizzard replied:
Pageserver constr is a new addition to etcd structure. So it is not already there from my point of view. 

Should we have a call then to discuss that? Because Option<String> allows only one pageserver, and the previous callmemaybe allowed to have more than one. 

Also do not get why it is a throwaway code

SomeoneToIgnore replied:
> Pageserver constr is a new addition to etcd structure.

But an old one to the safekeeper's logic, where they set this connstr to a timeline data when it's ready for streaming.
I am sure I'll be able to change that logic too "the right way", but don't want to touch it now, if possible.

> Should we have a call then to discuss that?

We can, but, honestly, I'm sick of this PR already and the pressure on its merge timings, better help me with the last failing test first 🙂 
In general, I see this particular thing as an improvement to the existing code, it can work without it.
Tests are crucial and block everything, on the other hand.

> Because Option allows only one pageserver, and the previous callmemaybe allowed to have more than one.

Not sure this is right: the SK code logic implies that a given SK Timeline has the same `Option` in its data and it was propagated this way by the old safekeepers to pageservers via callmemaybe.
Now, same option gets pushed by multiple safekeeprs (with possible multiple values), so nothing had really changed: a single Pageserver is able to select from multiple etcd connstrs the same way it was able to select from multiple callmemaybe connstrs before.

> Also do not get why it is a throwaway code

Mainly, since now we don't need to construct such string at all: AFAIK, it's needed to identify connecting pageservers and to indicate that the SK timeline is ready to be WAL streamed.
I use the latter, but the former could/should be done better via `NodeId` of the pageserver passed via etcd communications (as one of the ways).

LizardWizzard replied:
Hmmm, I think you're right. I'm reading it wrong. This is equal to pageserver that compute is connected to, so this should be ok in our current paradigm when tenant relocation requires compute restart. So after restart compute is connected to different PS and this connstring should be changed accordingly. 

> I use the latter, but the former could/should be done better via NodeId

Yeah, this case was one of the reasons to add node id to use it instead of connection string to identify pageserver.

Thanks!

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1289 @@ thread_local! {
+                    Some(sk_lsn_advantage) => {
+                        if sk_lsn_advantage >= self.max_lsn_wal_lag.get() {
+                            wal_lag_reason = Some(ReconnectReason::LaggingWal { current_lsn: Lsn(last_wal_receiver_data.ps_writelsn), new_lsn, threshold: self.max_lsn_wal_lag });
+                        }
```
We start from checking the lag, returning the check result as `wal_advantage_reconnect_reason`.
Later, we check the time and decide, which reason to return.

I wanted to do the "if Some" check only once hence a bit odd code, any better ideas on how to rewrite it better?

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1289 @@ thread_local! {
+    /// Returns a new candidate, if the current state is somewhat lagging, or `None` otherwise.
+    /// The current rules for approving new candidates:
+    /// * pick the safekeeper with biggest `commit_lsn` that's after than pageserver's latest Lsn for the timeline
+    /// * if the leader is the current SK and it has a different connection url — reconnect
```
Gave it a thought and indeed, if the connection URL change, the old one will break eventually, forcing the pageserver to disconnect form the old address and use the updated one from new etcd info.

SomeoneToIgnore commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -350,6 +433,9 @@ pub fn detach_timeline(
                 .detach_timeline(timeline_id)
                 .context("Failed to detach inmem tenant timeline")?;
             tenant.local_timelines.remove(&timeline_id);
+            tenants_state::try_send_timeline_update(LocalTimelineUpdate::Detach(
```
Feels like we'd better deal with both issues in https://github.com/neondatabase/neon/issues/1775 then?

LizardWizzard replied:
I'm ok with that, though the handling would be quite different. If I'm correct with my hypothesis for #1775 (I need to write a test to confirm that) it needs to take corresponding locks to ensure that compaction/gc are not running. And for the current problem here it requires some way to join the walreceiver which is a different story that this PR breaks.

Simpler option would be to pass a oneshot channel and let walrceiver send to it when it cleaned everything up. 

So up to you :) I'm ok with doing it in #1775 There shouldnt be any major problem from that because when data got deleted walreceiver will probably stop by itself because of the errors.

SomeoneToIgnore commented on pageserver/src/walreceiver/connection_handler.rs:
```diff
@@ -0,0 +1,398 @@
+                    // at risk of hitting a deadlock.
+                    ensure!(lsn.is_aligned());
+
+                    walingest.ingest_record(&timeline, recdata, lsn)?;
```
Now that I think of it, `timeline.tline.check_checkpoint_distance()?;` also calls some `.lock()` underneath, when spawning a flushing thread, for instance.
I've added a spawn_blocking around that, but it's hard to just wrap `walingest.ingest_record` into the same: it's used in two loops, so we have to catch almost every logic into this spawn.

Not sure how to do this properly without wrapping the walingest with `Arc<Mutex<..>>` which is also quite big.
Moveover, I did not find any locking inside that method, can you point me to that?

I'd leave it as is, if it's not causing issues, for now.

LizardWizzard replied:
I'm ok with leaving it as is, though maybe worth some comment. One lock is in timeline.writer() Which is called in modification.commit() in ingest record

SomeoneToIgnore commented:
I've added a multithreaded runtime with some arbitrary hadcoded value to address some basic concerns about performance, let's wait for load tests to see what's actually happening there.

The rest of the feedback is now addressed, one way or another.

hlinnaka commented on pageserver/src/walreceiver/connection_handler.rs:
```diff
@@ -0,0 +1,408 @@
+                }
+
+                let timeline_to_check = Arc::clone(&timeline.tline);
+                tokio::task::spawn_blocking(move || timeline_to_check.check_checkpoint_distance())
```
Why spawn_blocking here? Sure, check_checkpoint_distance() can block, but so can the ingest_record() call above.

The way I imagine this is that we use async to wait for a piece of WAL to arrive, across all the connections. But as soon as WAL arrives, the worker thread is responsible for handling it, and may block.

SomeoneToIgnore replied:
For mitigating the blocking effects exactly (the wrapped method's name is quite peculiar, tbh): https://github.com/neondatabase/neon/pull/1619#discussion_r881671949

I don't like the other blocking place too, but not sure how to fix that one simply now, so decided to leave as is per "ok" resolution in the discussion linked.

I'm not sure about the threads vs async tasks right now, since main part of that code is async and at this point it feels simpler to run one big `spawn_blocking` around the whole thing and stay within main async paradigm generally.

But I would like to see some load tests' numbers first, before guessing and changing this all.

hlinnaka commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1148 @@ thread_local! {
-// This is the entry point for the WAL receiver thread.
-//
-fn thread_main(conf: &'static PageServerConf, tenant_id: ZTenantId, timeline_id: ZTimelineId) {
-    let _enter = info_span!("WAL receiver", timeline = %timeline_id, tenant = %tenant_id).entered();
```
This `info_span!` it pretty important for logging. It makes sure that for every error that happens when processing the incoming WAL has the timeline and tenant ID attached to it. I don't see anything to replace this.

SomeoneToIgnore replied:
There's a bunch of others now, used as `.instrument(info_span` on a high-level futures such as `tokio::spawn` and such to replace this.

So far, the new span coverage looks good to me, here's an example: [test_pageserver_catchup logs](https://circleci-tasks-prod.s3.us-east-1.amazonaws.com/storage/artifacts/2d7b9e85-192b-4163-bc6c-303d0b8bb7d2/19954173-edff-40d2-9cce-53af749dd347/0/tmp/test_output/test_pageserver_catchup_while_compute_down/repo/pageserver.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQVFQINEOEPKIPAUE%2F20220526%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220526T084211Z&X-Amz-Expires=60&X-Amz-Security-Token=IQoJb3JpZ2luX2VjELn%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIQCyk41ON%2Bx5qCh4clO0BzuWSS6GC46UJgkkKNJQLs27mgIgLn0QL79D%2BDjxvGRZf%2FwGNGrKo6UPCbODmfZ7nXyNalIqtAIIov%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARACGgwwNDU0NjY4MDY1NTYiDLciCOG8xPuNf6M4QiqIAqJBb%2FVVyHW%2Ft9qNk1AOQ%2Bz7DAmK0QrlDvxsZ26LvaPfaQc6Vku2fefSjtISC7qBowItoM9k%2FJFW%2BgNob76mgpyR4k50B8kyAfkSTkDFZTo3MpjN9Vxx67fwhCpYNAPTSVEhArW5eDbeNswxntTNS7SiC2mAKrmKOxQVMSgU0mRCroTk60yFRuD0rEuvOggjsLyZi0xu5n%2FgYTq3pqBBHq4G2eFFXfcuS8Mj9xcdkfER2Y%2FBpBG%2FA4D8cBTR6audm7n34S0vfETPZvaaJOIA8FcCeVBmdVuECIarYSjh0o3PyApZtDsiNKasP1Zq3p2A6YJhGvRFEdpgUg4PekX1zobVfXN0xskhGzDg%2BryUBjqdAZuJ0f0rti8vGxkHIrCoqUAzAvNagkV3NflqMK%2FJ4GMdP4MTvYTA5LOPDNqAIFww3dsB2EANy7o0DzUgZw2wVadOZYnCwk7Noeimr1yHe%2Fm8aBs4QpivGj%2BAeb2yiu%2FVMbqdwxyZYX3jJkWyjyjrp8J8k9%2ByLdr1zkOVl5ccTv2BNtYsH2eAaRjW60UGGtPSVB2vUoLKigwV1n21WS4%3D&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=397172ba7521750c8a9d25a63ca13fff5927a37acf890b11b34c0aba626fc0ef)
(grep by `wal_receiver_main`)

Not sure about span names and data placement, but looks ok-ish for the first attempt too, IMO.

SomeoneToIgnore commented:
Thanks to @LizardWizzard , I've got a good way to repro the issue: it was related to `/detach` call that was not tested in our e2e tests really.
I've pushed a separate commit in case someone is interested to see what's changed, but plan to merge it with the first commit before pushing.

Now all the tests are green.

arssher commented on safekeeper/src/receive_wal.rs:
```diff
@@ -85,6 +85,14 @@ impl<'pg> ReceiveWalConn<'pg> {
             _ => bail!("unexpected message {:?} instead of greeting", next_msg),
         }
 
+        // Register the connection and defer unregister.
```
We do this below.

SomeoneToIgnore replied:
Merge artifact, removed, thanks.

LizardWizzard commented on docs/settings.md:
```diff
@@ -111,6 +111,21 @@ L0 delta layer threshold for L1 iamge layer creation. Default is 3.
+
+Difference between Lsn values of the latest available WAL on safekeepers: if currently connected safekeeper starts to lag too long and too much,
+it gets swapped to the different one.
+A safekeeper has to be both lagging beyond the threshold and do that over `lagging_wal_timeout` time in order to be considered as lagging.
```
@arssher WDYT?

arssher replied:
I think these conditions should be ORed, and ANDed -- either of them is trigger to switch to more advanced safekeeper.
The idea is, relying only on timeout is not enough because safekeeper might lag a lot and slowly catch up. Relying only on lag is not enough because we might be in deadlock when new WAL is not generated because old hasn't reached pageserver. However, both values must be big enough to prevent flip-flopping during normal work.

arssher replied:
... and looking at the code, these are ORed actually, so just doc needs a fix.

SomeoneToIgnore replied:
Actually, the code has both conditions to check also, but apparently it's written a bit complicated, since fooled a second person already.

I've got your point and will apply OR everywhere.

arssher commented on safekeeper/src/timeline.rs:
```diff
@@ -191,79 +190,33 @@ impl SharedState {
-    /// start/change walsender (via callmemaybe).
-    fn callmemaybe_sub(
+    /// Activate timeline's walsender: start/change timeline information propagated into etcd for further pageserver connections.
+    fn activate_walsender(
```
This can be removed altogether, with new model don't need anything on safekeeper side except for pushing to etcd.

SomeoneToIgnore replied:
I would leave it right now in this PR, since there's some logic in safekeeper, related to stopping walsender via setting this field to `None` in the timeline (hence, it stops sending it to etcd and pageserver does not see it in the list of candidates): https://github.com/neondatabase/neon/blob/0e1bd57c533165dbe4bead8fa23baefa09c97b82/safekeeper/src/send_wal.rs#L318

I agree that etcd subscription should be removed here instead, but let's do that separately.

SomeoneToIgnore replied:
https://github.com/neondatabase/neon/issues/1844

arssher commented:
Hasn't looked through the all code yet, but generally looks good (and much better than on prev iteration) and I don't mind merging, there was quite some eyeballing already.

SomeoneToIgnore commented:
Looks like there are 4 leftovers to do after this PR:

* remove `callmemaybe_connstring` from the compute code (will do next)
* properly test the performance implications of this PR: will implicitly do after the merge with everybody's help, will eventually explicitly do with load testing later. 
Will require considerations on the `tokio::task` + `spawn_blocking` vs `std::thread` approach towards the WAL streaming, but for that we need to be able to compare load test results over some metrics.
* `pageserver_connstr` has to be removed entirely: https://github.com/neondatabase/neon/issues/1844
* "safe" races with detach: https://github.com/neondatabase/neon/issues/1775#issuecomment-1143511293

LizardWizzard commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -286,21 +359,31 @@ pub fn activate_tenant(tenant_id: ZTenantId) -> anyhow::Result<()> {
-
-        TenantState::Broken => {
-            // cannot activate
+        (TenantState::Active, TenantState::Stopping | TenantState::Idle) => {
```
FYI @bojanserafimov

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1133 @@ thread_local! {
+                        ) => {},
+                    }
+                }
+            }.instrument(info_span!("wal_receiver_main")));
```
BTW, this span in inherited by all subsequent tasks, which is a bit confusing (they are not main). Do you know a way to clear spans for them?

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1133 @@ thread_local! {
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("wal-receiver-runtime-thread")
+        .worker_threads(40)
```
Don't think this requires immediate action, but I'm still dubious about running walreceivers in general tokio runtime. I'd expect them to be not idle-waiting-for-network most of the time, in which case async runtime is pointless. Moreover, operations like locking layer while writing to it are not async friendly and will block the runtime with etcd communcation and main wal receiver activity.

LizardWizzard replied:
>  in which case async runtime is pointless.

That's simply not true. Async is useful not only in places where most of the tasks are idle. E g for cancellation purposes we will need to use per-thread runtimes in each walreceiver which IMO is not ideal because one of the main reasons for introducing async paradigm was to decrease the overhead for the separate execution thread so one can have thousands of coroutines running concurrently (and some of them simultaneously). Per thread runtimes make sense when you want to reduce cross core coordination and increase thread locality. It works via spawning threads for every available processor core and using a separate runtime in every thread. This is the paradigm that is in use by [Actix](https://actix.rs/) or nginx. Secondly, aside from cancellation semantics async is capable of introducing higher parallelization via adding more concurrency to io operations. E g you can schedule many write operations simultaneously and wait for them at once instead of issuing them one by one. Not to mention io_uring capabilities.

Also, there are some attempts to reduce thread count: https://github.com/neondatabase/neon/pull/1794, and IMO having per thread wal_receivers contradicts this direction.

Also, there is nothing wrong with using async for more CPU-intensive tasks. See the article: https://www.influxdata.com/blog/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/ and write up on the subject in the repository https://github.com/influxdata/influxdb_iox/blob/main/docs/multi_core_tasks.md

Blocking is a problem and it partially can be solved by spawn_blocking, but it uses threadpool under the hood so a bit of a compromise.

To be clear, I do not insist on rewriting everything to async. It has its own problems. But let's be fair in accounting for its benefits too.

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1133 @@ thread_local! {
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("wal-receiver-runtime-thread")
+        .worker_threads(40)
```
Don't think this requires immediate action, but I'm still dubious about running walreceivers in general tokio runtime. I'd expect them to be not idle-waiting-for-network most of the time, in which case async runtime is pointless. Moreover, operations like locking layer while writing to it are not async friendly and will block the runtime with etcd communcation and main wal receiver activity.

arssher commented on pageserver/src/walreceiver.rs:
```diff
@@ -64,375 +80,1133 @@ thread_local! {
+    }
 
-                waldecoder.feed_bytes(data);
+    info!("Waiting for the current connection to close");
```
BTW, closing connection here means even on short etcd outage we immediately lose current walreceiver connection. We could keep it.
